src/main/java/com/commercetools/sync/services/CustomerService.java

```diff
@@ -0,0 +1,123 @@
+     * potentially injected {@link SphereClient}. If there is no matching resource an empty {@link Optional} will be
+     * returned in the returned future. A mapping of the key to the id of the fetched customer is persisted in an in
+     * -memory map.
+     *
+     * @param key the key of the resource to fetch
+     * @return {@link CompletionStage}&lt;{@link Optional}&gt; in which the result of it's completion contains an {@link
+     *         Optional} that contains the matching {@link Customer} if exists, otherwise empty.
+     */
+    @Nonnull
+    CompletionStage<Optional<Customer>> fetchCustomerByKey(@Nullable String key);
+
+    /**
+     * Given a {@code key}, if it is blank (null/empty), a completed future with an empty optional is returned.
+     * Otherwise this method checks if the cached map of resource keys -&gt; ids contains the given key. If it does, an
+     * optional containing the matching id is returned. If the cache doesn't contain the key; this method attempts to
+     * fetch the id of the key from the CTP project, caches it and returns a {@link CompletionStage}&lt; {@link
+     * Optional}&lt;{@link String}&gt;&gt; in which the result of it's completion could contain an {@link Optional}
+     * holding the id or an empty {@link Optional} if no customer was found in the CTP project with this key.
+     *
+     * @param key the key by which a customer id should be fetched from the CTP project.
+     * @return {@link CompletionStage}&lt;{@link Optional}&lt;{@link String}&gt;&gt; in which the result of it's
+     *         completion could contain an {@link Optional} holding the id or an empty {@link Optional} if no customer
+     *         was found in the CTP project with this key.
+     */
+    @Nonnull
+    CompletionStage<Optional<String>> fetchCachedCustomerId(@Nonnull String key);
+
+    /**
+     * Given a resource draft of type {@link CustomerDraft}, this method attempts to create a resource {@link Customer}
+     * based on the draft, in the CTP project defined by the sync options.
+     *
+     * <p>A completion stage containing an empty option and the error callback will be triggered in those cases:
+     * <ul>
+     *     <li>the draft has a blank key</li>
+     *     <li>the create request fails on CTP</li>
+     * </ul>
+     *
+     * <p>On the other hand, if the resource gets created successfully on CTP, the created resource's id and
+     * key are cached and the method returns a {@link CompletionStage} in which the result of it's completion
+     * contains an instance {@link Optional} of the resource which was created.
+     *
+     * @param customerDraft the resource draft to create a resource based off of.
+     * @return a {@link CompletionStage} containing an optional with the created resource if successful otherwise an
+     *         empty optional.
+     */
+    @Nonnull
+    CompletionStage<Optional<Customer>> createCustomer(@Nonnull CustomerDraft customerDraft);
```

Comment:
There's also create customer in a store. Do we need to consider that as well? https://docs.commercetools.com/http-api-projects-customers#create-customer-sign-up-in-a-store-beta

Reply:
haven't considered it yet because its in a beta state. @ahmetoz do we need to consider that?

Reply:
We will do the syncing for customers not store customers. Some endpoints in customer is more related to the UI usages. (/{projectKey}/in-store/key={storeKey}/customers) as you could see the endpoint is not directly a customer endpoint like (/{projectKey}/customers/{id})